### PR TITLE
Delay possible inference processing in strings

### PR DESCRIPTION
**Pull request by ajreynol**
This is work towards a model-based variant of the core strings solver.

This moves lemma processing to a more global place in the normal forms inference lemma schema in preparation for invoking the model-based module there.

### CODE REVIEW

File: src/theory/strings/core_solver.cpp [hunk 1]
```diff
@@ -529,17 +529,25 @@ void CoreSolver::checkNormalFormsEq()
   std::map<Node, Node> nf_to_eqc;
   std::map<Node, Node> eqc_to_nf;
   std::map<Node, Node> eqc_to_exp;
+  std::vector<CoreInferInfo> pinfer;
```
Comment (yoni206) [inline src/theory/strings/core_solver.cpp:532]: I know none of these local variables is documented, but perhaps at least this one can be?

File: src/theory/strings/core_solver.cpp [hunk 2]
```diff
@@ -619,8 +635,8 @@ void CoreSolver::normalizeEquivalenceClass(Node eqc, TypeNode stype)
       return;
     }
     // process the normal forms
-    processNEqc(eqc, normal_forms, stype);
-    if (d_im.hasProcessed())
+    processNEqc(eqc, normal_forms, stype, pinfer);
+    if (d_im.hasProcessed() || !pinfer.empty())
```
Comment (yoni206) [inline src/theory/strings/core_solver.cpp:639]: Can you add a comment in the code saying why do we return in this case?

File: src/theory/strings/core_solver.cpp [hunk 3]
```diff
@@ -2656,6 +2637,39 @@ void CoreSolver::checkLengthsEqc() {
   }
 }
 
+void CoreSolver::processPossibleInferInfo(std::vector<CoreInferInfo>& pinfer)
+{
+  // now, determine which of the possible inferences we want to add
```
Comment (yoni206) [inline src/theory/strings/core_solver.cpp:2642]:
```suggestion
  // determine which of the possible inferences we want to add
```

File: src/theory/strings/core_solver.cpp [hunk 4]
```diff
@@ -529,17 +529,25 @@ void CoreSolver::checkNormalFormsEq()
...
-    normalizeEquivalenceClass(eqc, stype);
+    normalizeEquivalenceClass(eqc, stype, pinfer);
     Trace("strings-debug") << "Finished normalizing eqc..." << std::endl;
     if (d_im.hasProcessed())
     {
       return;
     }
+    if (!pinfer.empty())
+    {
+      // if we had a possible inference, we were unable to assign
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/strings/core_solver.cpp:547]: This is the contract of `normalizeEquivalenceClass`, I'll add to its doc.

File: src/theory/strings/core_solver.cpp (hunk 1 above)
Comment (ajreynol) [reply src/theory/strings/core_solver.cpp:540]: Now documented.

**Review by yoni206 — APPROVED** [review]
Thanks for the changes.

**Pull request merged**